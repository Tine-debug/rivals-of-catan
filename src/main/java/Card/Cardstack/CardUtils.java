package Card.Cardstack;

import java.util.ArrayList;
import java.util.List;

import Card.Card;

public class CardUtils {

    public static Card popCardByName(List<Card> cards, String name) {
        if (cards == null || name == null) {
            return null;
        }
        String target = name.trim();
        for (int i = 0; i < cards.size(); i++) {
            Card c = cards.get(i);
            if (c != null && c.getName() != null && c.getName().trim().equalsIgnoreCase(target)) {
                return cards.remove(i);
            }
        }
        return null;
    }

    public static List<Card> extractBy(List<Card> cards, String attribute, String value) {
        List<Card> out = new ArrayList<>();
        for (int i = cards.size() - 1; i >= 0; i--) {
            Card c = cards.get(i);
            switch (attribute) {
                case "type":
                    if (c.getType().equals(value)) {
                        out.add(0, cards.remove(i));
                    }
                    break;
                case "name":
                    if (c.getName().equals(value)) {
                        out.add(0, cards.remove(i));
                    }
                    break;
                case "theme":
                    if (c.getTheme().equals(value)) {
                        out.add(0, cards.remove(i));
                    }
                    break;
                case "placement":
                    if (c.getPlacement().equals(value)) {
                        out.add(0, cards.remove(i));
                    }
                    break;
                default:
                    // unknown attribute, return what we got
                    return out;
            }
        }
        return out;
    }
}
