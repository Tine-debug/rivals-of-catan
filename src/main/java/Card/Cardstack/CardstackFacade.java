package Card.Cardstack;

import Player.Player;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import Card.Card;

public class CardstackFacade {

    private final RegionManager regionManager;
    private final CenterBuildingManager buildingManager;
    private final EventManager eventManager;
    private final DrawStackManager drawStackManager;
    private final CardUtils cardUtils;

    private CardstackFacade() {
        this.regionManager = new RegionManager();
        this.buildingManager = new CenterBuildingManager();
        this.eventManager = new EventManager();
        this.drawStackManager = new DrawStackManager();
        this.cardUtils = new CardUtils();
    }

    static private CardstackFacade instance = null;

    static public CardstackFacade getInstance() {
        if (instance == null) {
            instance = new CardstackFacade();
        }
        return instance;
    }

    public void loadCards(String jsonPath, String theme) throws IOException{
        List<Card> all = LoadCards.loadCards(jsonPath, true, theme);
        int allsize = all.size();

        buildingManager.initializeCenterCards(all);
        regionManager.initializeRegions(all);
        eventManager.initializeEvents(all, true);
        drawStackManager.initializeStacks(all, true, allsize);

    }

    public List<Card> getStack(String Name) {
        return this.buildingManager.getStack(Name);
    }

    public List<Card> loadCardsForTesting(String jsonPath, String desiredTheme, boolean shuffle, boolean loadmulitpe) throws IOException {
        List<Card> all = LoadCards.loadCards(jsonPath, loadmulitpe, "basic");
        List<Card> out = new ArrayList<>(all);
        buildingManager.initializeCenterCards(all);
        regionManager.initializeRegions(all);
        eventManager.initializeEvents(all, shuffle);
        drawStackManager.initializeStacks(all, shuffle, out.size());
        return out;
    }

    public void initializePlayerPrincipality(Player p, int[][] dice, int center, int idx) {
        this.buildingManager.inizializePrincipiality(p, dice, center, idx, regionManager.getRegions());
    }

    public void shuffleRegions() {
        regionManager.shuffle();
    }

    public Card drawRegionCard() {
        return regionManager.draw();
    }

    public int getRegionStackSize() {
        return regionManager.size();
    }

    public String printRegionStack() {
        return regionManager.print();
    }

    public void resetEventStack() {
        eventManager.reset();
    }

    public Card drawEventCard() {
        return eventManager.draw();
    }

    public void placeCardBottomStack(Card c, int stackNum) {
        drawStackManager.placeAtBottom(c, stackNum);
    }

    public void drawCardFromStack(int stack, Player p) {
        drawStackManager.drawCard(stack, p);
    }

    public void chooseCardFromStack(int stack, Player p) {
        drawStackManager.chooseCard(stack, p);
    }

    public void drawCard(int stack, Player p) {
        drawStackManager.drawCard(stack, p);
    }

    public int[] placeCenterCard(Player active, Player other, String cardType) {
        return buildingManager.placeCenterCard(active, other, cardType);
    }

    public List<String> getCenterBuildingCosts() {
        return buildingManager.getBuildingCosts();
    }

    public Card pickRegionByNameOrIndex(String spec) {
        return regionManager.pickByNameOrIndex(spec);
    }

    public Card findUndicedRegionByName(String name) {
        return regionManager.findUndiced(name);
    }

    public Card popCardByName(List<Card> cards, String name) {
        return CardUtils.popCardByName(cards, name);
    }

}
