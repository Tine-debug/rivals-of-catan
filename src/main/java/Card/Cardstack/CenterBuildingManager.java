package Card.Cardstack;

import Card.Card;
import Player.Player;
import java.util.*;

public class CenterBuildingManager {

    private List<Card> roads = new ArrayList<>();
    private List<Card> settlements = new ArrayList<>();
    private List<Card> cities = new ArrayList<>();


    public void initializeCenterCards(List<Card> allCards) {
        roads = CardUtils.extractBy(allCards, "name", "Road");
        settlements = CardUtils.extractBy(allCards, "name", "Settlement");
        cities = CardUtils.extractBy(allCards, "name", "City");
    }

    public List<String> getBuildingCosts() {
        List<String> costs = new ArrayList<>();
        if (!roads.isEmpty()) costs.add("ROAD(" + getCost(roads.get(0)) + ")");
        if (!settlements.isEmpty()) costs.add("SETTLEMENT(" + getCost(settlements.get(0)) + ")");
        if (!cities.isEmpty()) costs.add("CITY(" + getCost(cities.get(0)) + ")");
        return costs;
    }

    private String getCost(Card card) {
        return card.getCost() == null ? "-" : card.getCost();
    }

    public List<Card> getStack(String Name){
        switch (Name) {
            case "Road": return roads;
            case "Settlement": return settlements;
            case "City": return cities;
            default:
                return null;
        }
    }

    public int[] placeCenterCard(Player active, Player other, String type) {
 List<Card> pile = null;
        if (type.equalsIgnoreCase("Road")) {
            pile = roads;
        } else if (type.equalsIgnoreCase("Settlement")) {
            pile = settlements;
        } else if (type.equalsIgnoreCase("City")) {
            pile = cities;
        }

        if (pile == null || pile.isEmpty()) {
            active.sendMessage("No " + type + " cards left in the pile.");
            return new int[]{-1, -1};
        }

        Card proto = pile.get(0);

        if (!active.payCost(proto.getCost())) {
            active.sendMessage("Can't afford cost: " + (proto.getCost() == null ? "-" : proto.getCost()));
            return new int[]{-1, -1};
        }

        active.sendMessage("PROMPT: Enter placement coordinates as: ROW COL");
        int row, col;
        try {
            String[] rc = active.receiveMessage().trim().split("\\s+");
            row = Integer.parseInt(rc[0]);
            col = Integer.parseInt(rc[1]);
        } catch (NumberFormatException e) {
            active.sendMessage("Invalid coordinates. Use: ROW COL (e.g., 2 3)");
            active.refundCost(proto.getCost());
            return new int[]{-1, -1};
        }

        boolean ok = proto.applyEffect(active, other, row, col);
        if (!ok) {
            active.sendMessage("Illegal placement/effect; refunding cost.");
            active.refundCost(proto.getCost());
            return new int[]{-1, -1};
        }

        pile.remove(0);
        return new int[]{row, col};

    }


    public void inizializePrincipiality(Player p, int[][] dice, int center, int i, List<Card> regions) {
        p.placeCard(center, 1, CardUtils.popCardByName(settlements, "Settlement"));
        p.placeCard(center, 2, CardUtils.popCardByName(roads, "Road"));
        p.placeCard(center, 3, CardUtils.popCardByName(settlements, "Settlement"));

        String[] names = {"Forest", "Gold Field", "Field", "Hill", "Pasture", "Mountain"};
        for (int j = 0; j < names.length; j++) {
            Card card = CardUtils.popCardByName(regions, names[j]);
            if (card != null) {
                card.setDiceRoll(dice[i][j]);
                card.setRegionProduction(names[j].equals("Gold Field") ? 0 : 1);
            }
            int row = j < 3 ? center - 1 : center + 1;
            int col = (j % 3) * 2;
            p.placeCard(row, col, card);
        }
    }
}
