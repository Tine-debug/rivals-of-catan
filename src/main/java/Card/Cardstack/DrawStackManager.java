package Card.Cardstack;

import Card.Card;
import Player.Player;

import java.util.*;
import java.io.IOException;

public class DrawStackManager {

    private List<Card> drawStack1 = new ArrayList<>();
    private List<Card> drawStack2 = new ArrayList<>();
    private List<Card> drawStack3 = new ArrayList<>();
    private List<Card> drawStack4 = new ArrayList<>();

    public void initializeStacks(List<Card> allCards, boolean shuffle, int size) throws IOException{
        int stackSize = 9;
        if (shuffle) Collections.shuffle(allCards);
        if (allCards.size() < 1) {
            throw new IOException("No cards loaded into drawstacks. All Size: " + size);
        }

        drawStack1 = new ArrayList<>(allCards.subList(0, Math.min(stackSize, allCards.size())));
        drawStack2 = new ArrayList<>(allCards.subList(Math.min(stackSize, allCards.size()),
                Math.min(2 * stackSize, allCards.size())));
        drawStack3 = new ArrayList<>(allCards.subList(Math.min(2 * stackSize, allCards.size()),
                Math.min(3 * stackSize, allCards.size())));
        drawStack4 = new ArrayList<>(allCards.subList(Math.min(3 * stackSize, allCards.size()),
                Math.min(4 * stackSize, allCards.size())));
    }

    private List<Card> stackBy(int n) {
        switch (n) {
            case 1:
                return drawStack1;
            case 2:
                return drawStack2;
            case 3:
                return drawStack3;
            case 4:
                return drawStack4;
            default:
                return drawStack1;
        }
    }

    public void placeAtBottom(Card card, int stackNum) {
        if (card == null) {
            return;
        }
        List<Card> stack = stackBy(stackNum);
        stack.add(card);
    }

    public void drawCard(int which, Player p) {
        List<Card> stack = stackBy(which);
        if (stack.isEmpty()) {
            p.sendMessage("That stack is empty.");
            return;
        }
        p.addToHand(stack.remove(0));
    }

    public void drawCardFromStack(int which, Player p) {
        List<Card> stack = stackBy(which);
        if (stack.isEmpty()) {
            int tries = 0;
            do {
                which = 1 + (which % 4);
                stack = stackBy(which);
                tries++;
            } while (stack.isEmpty() && tries <= 4);

            if (stack.isEmpty()) {
                p.sendMessage("All stacks empty.");
                return;
            }
        }
        p.addToHand(stack.remove(0));
    }

    public void chooseCard(int chosen, Player p) {
        List<Card> stack = stackBy(chosen);
        if (stack.isEmpty()) {
            p.sendMessage("That stack is empty.");
            return;
        }
        p.sendMessage("Stack contains (top..bottom):");
        for (Card c : stack) {
            p.sendMessage(" - " + c.toString());
        }
        p.sendMessage("PROMPT: Type exact name to take:");
        String take = p.receiveMessage();
        for (int i = 0; i < stack.size(); i++) {
            if (stack.get(i).getName().equalsIgnoreCase(take)) {
                p.addToHand(stack.remove(i));
                return;
            }
        }
        p.sendMessage("Not found; no card taken.");
    }
}
