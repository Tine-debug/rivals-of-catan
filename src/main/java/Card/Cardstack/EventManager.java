package Card.Cardstack;

import Card.Card;

import java.util.*;

public class EventManager {

    private List<Card> events = new ArrayList<>();
    private List<Card> resolved = new ArrayList<>();

    public void initializeEvents(List<Card> allCards, boolean shuffle) {
        events = CardUtils.extractBy(allCards, "placement", "Event");
        Card yule = CardUtils.popCardByName(events, "Yule");
        if (shuffle) Collections.shuffle(events);
        if (yule != null && events.size() >= 3) {
            events.add(Math.max(0, events.size() - 3), yule);
        }
    }

    public void reset() {
        events.addAll(resolved);
        resolved.clear();
        Card yule = CardUtils.popCardByName(events, "Yule");
        Collections.shuffle(events);
        if (yule != null && events.size() >= 3) {
            events.add(Math.max(0, events.size() - 3), yule);
        }
    }

    public Card draw() {
        if (events.isEmpty()) return null;
        Card c = events.remove(0);
        resolved.add(c);
        return c;
    }
}
