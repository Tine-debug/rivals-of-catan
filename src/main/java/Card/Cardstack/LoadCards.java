package Card.Cardstack;

import Card.Card;
import Card.Cardbuilder;

import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import Points.PointsBuilder;
import Card.logic.LogicFactory;
import java.io.IOException;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

public class LoadCards {

    public static ArrayList<Card> loadCards(String jsonPath, boolean loadmultiple, String desiredTheme) throws IOException {

        ArrayList<Card> allBasic = new ArrayList<>();

        try (FileReader fr = new FileReader(jsonPath)) {
            JsonElement root = JsonParser.parseReader(fr);
            if (!root.isJsonArray()) {
                throw new IOException("cards.json: expected top-level array");
            }
            JsonArray arr = root.getAsJsonArray();

            for (JsonElement el : arr) {
                if (!el.isJsonObject()) {
                    continue;
                }
                JsonObject o = el.getAsJsonObject();
                String theme = gs(o, "theme");
                if (theme == null || !theme.toLowerCase().contains(desiredTheme.toLowerCase())) {
                    continue;
                }
                int number;
                if (loadmultiple) {
                    number = gi(o, "number", 1);
                } else {
                    number = 1;
                }
                for (int i = 0; i < number; i++) {
                    Cardbuilder cardbuilder = new Cardbuilder();
                    String name = gs(o, "name");
                    cardbuilder.name(name);
                    String type = gs(o, "type");
                    cardbuilder.type(type);
                    String placement = gs(o, "placement");
                    cardbuilder.placement(placement);
                    cardbuilder.logic(LogicFactory.createLogic(placement, name, type));

                    cardbuilder.oneOf(gs(o, "oneOf"));
                    cardbuilder.cardText(gs(o, "cardText"));
                    cardbuilder.cost(gs(o, "cost"));

                    PointsBuilder pointsBuilder = new PointsBuilder();
                    pointsBuilder.victoryPoints(gs(o, "victoryPoints"));
                    pointsBuilder.commercePoints(gs(o, "CP"));
                    pointsBuilder.strengthPoints(gs(o, "SP"));
                    pointsBuilder.skillPoints(gs(o, "FP"));
                    pointsBuilder.progressPoints(gs(o, "PP"));
                    pointsBuilder.sailPoints(gs(o, "LP"));
                    pointsBuilder.canonPoints(gs(o, "KP"));

                    cardbuilder.points(pointsBuilder.build());

                    allBasic.add(cardbuilder.build());
                }
            }
        }

        return allBasic;

    }

    private static String gs(JsonObject o, String k) {
        if (!o.has(k)) {
            return null;
        }
        JsonElement e = o.get(k);
        return (e == null || e.isJsonNull()) ? null : e.getAsString();
    }

    private static int gi(JsonObject o, String k, int def) {
        if (!o.has(k)) {
            return def;
        }
        try {
            return o.get(k).getAsInt();
        } catch (Exception e) {
            return def;
        }
    }
}
