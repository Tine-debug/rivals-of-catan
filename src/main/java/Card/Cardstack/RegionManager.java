package Card.Cardstack;

import Card.Card;

import java.util.*;

public class RegionManager {
    private List<Card> regions = new ArrayList<>();

    public void initializeRegions(List<Card> cards) {
        regions = CardUtils.extractBy(cards, "type", "Region");
    }

    public void shuffle() {
        Collections.shuffle(regions);
    }

    public Card draw() {
        return regions.isEmpty() ? null : regions.remove(0);
    }

    public int size() {
        return regions.size();
    }

    public String print() {
        String regionsnames = "";
        for (Card card : regions) {
            regionsnames += card.toString() + "\n";
        }
        return regionsnames;
    }

    public Card pickByNameOrIndex(String spec) {
        if (spec == null || spec.isBlank()) return null;
        try {
            int idx = Integer.parseInt(spec.trim());
            return idx >= 0 && idx < regions.size() ? regions.remove(idx) : null;
        } catch (NumberFormatException ignored) {
        }
        return CardUtils.popCardByName(regions, spec);
    }

    public Card findUndiced(String name) {
        return regions.stream()
                .filter(c -> name.equalsIgnoreCase(c.getName()) && c.getdiceRoll() == 0)
                .findFirst().orElse(null);
    }

    public List<Card> getRegions(){
        return regions;
    }

}
