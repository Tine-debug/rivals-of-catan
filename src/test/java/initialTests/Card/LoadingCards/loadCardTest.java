
import java.io.OutputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;
import java.util.stream.Stream;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;

import Card.Cardstack.CardstackFacade;

public class loadCardTest {

    public final CardstackFacade stacks = CardstackFacade.getInstance();

    @Test
    public void loadCardtest() {
        try {
            stacks.loadCardsForTesting("cards.json", "basic", false, true);
        } catch (Exception e) {
            System.err.println("Failed to load");
        }
        Approvals.verify(stacks.printRegionStack());

    }
    

}
