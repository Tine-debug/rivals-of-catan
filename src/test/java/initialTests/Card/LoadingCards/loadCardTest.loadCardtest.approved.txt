Field
Field
Field
Field
Mountain
Mountain
Mountain
Mountain
Gold Field
Gold Field
Gold Field
Gold Field
Hill
Hill
Hill
Hill
Forest
Forest
Forest
Forest
Pasture
Pasture
Pasture
Pasture
